### PR TITLE
chore(middleware/cors): Merge changes from v2

### PR DESCRIPTION
**Pull request by sixcolors**
Merges CORS middleware changes from v2 branch to main

Pending merge of #2921 

## Summary by CodeRabbit

- **New Features**
	- Enhanced CORS middleware functionality with dynamic origin validation, subdomain matching, and improved configuration options.
- **Bug Fixes**
	- Refined handling of allowed origins, updated header settings, and fixed issues with origin processing for improved security.
- **Documentation**
	- Added detailed explanations of CORS concepts, security considerations, and configuration options.
- **Tests**
	- Expanded test coverage for different origins and request methods to ensure proper CORS header handling.

### CODE REVIEW

File: middleware/cors/cors.go [hunk 1]
```diff
@@ -176,57 +212,65 @@
...
-			} else if allowOrigin == "*" {
-				log.Warn("[CORS] 'AllowCredentials' is true. Ensure 'AllowOrigins' is not set to '*' in the configuration.")
-			}
-		} else {
-			// For non-credential requests, it's safe to set to '*' or specific origins
+		setCORSHeaders(c, allowOrigin, allowMethods, allowHeaders, exposeHeaders, maxAge, cfg)
+
+		// Send 204 No Content
+		return c.SendStatus(fiber.StatusNoContent)
+	}
```
[the remark this answers is not in the document]
Comment (sixcolors) [reply middleware/cors/cors.go:229]: Can you give me a committable suggestion?

File: middleware/cors/cors.go [hunk 2]
```diff
@@ -110,31 +110,48 @@
...
+					continue
+				}
+				sd := subdomain{prefix: normalizedOrigin[:i+3], suffix: normalizedOrigin[i+3:]}
+				allowSOrigins = append(allowSOrigins, sd)
+			} else {
+				normalizedOrigin, isValid := processOrigin(origin)
+				if !isValid {
+					continue
+				}
+				allowOrigins = append(allowOrigins, normalizedOrigin)
```
[the remark this answers is not in the document]
Comment (sixcolors) [reply middleware/cors/cors.go:150]: Check cors_test.go and verify if coverage is already there for allowed origins.

File: middleware/cors/utils_test.go [hunk 3]
```diff
@@ -75,44 +75,6 @@ func Test_matchScheme(t *testing.T) {
...
-
-		if result != tc.expected {
-			t.Errorf("Expected validateOrigin('%s', '%s') to be %v, but got %v", tc.domain, tc.pattern, tc.expected, result)
-		}
-	}
-}
-
 // go test -run -v Test_normalizeDomain
 func Test_normalizeDomain(t *testing.T) {
 	testCases := []struct {
```
[the remark this answers is not in the document]
Comment (sixcolors) [reply middleware/cors/utils_test.go:80]: Sure

File: middleware/cors/utils.go [hunk 4]
```diff
@@ -83,3 +59,13 @@ func normalizeOrigin(origin string) (bool, string) {
...
+
+type subdomain struct {
+	// The wildcard pattern
+	prefix string
+	suffix string
+}
+
+func (s subdomain) match(o string) bool {
+	return len(o) >= len(s.prefix)+len(s.suffix) && strings.HasPrefix(o, s.prefix) && strings.HasSuffix(o, s.suffix)
+}
```
[the remark this answers is not in the document]
Comment (sixcolors) [reply middleware/cors/utils.go:71]: Sounds good

File: docs/api/middleware/cors.md [hunk 5]
```diff
@@ -4,13 +4,15 @@ id: cors
...
+The middleware uses the `AllowOrigins` option to control which origins can make cross-origin requests. It supports single origin, multiple origins, subdomain matching, and wildcard origin. It also allows programmatic origin validation with the `AllowOriginsFunc` option.
 
-When defining your Origins make sure they are properly formatted. The middleware validates and normalizes the provided origins, ensuring they're in the correct format by checking for valid schemes (http or https), and removing any trailing slashes.
+To ensure that the provided `AllowOrigins` origins are correctly formatted, this middleware validates and normalizes them. It checks for valid schemes, i.e., HTTP or HTTPS, and it will automatically remove trailing slashes. If the provided origin is invalid, the middleware will panic.
+
+When configuring CORS, it's important to avoid [common pitfalls](#common-pitfalls) like using a wildcard origin with credentials, being overly permissive with origins, and inadequate validation with `AllowOriginsFunc`. Misconfiguration can expose your application to various security risks.
 
 ## Signatures
```
[the remark this answers is not in the document]
Comment (sixcolors) [reply docs/api/middleware/cors.md:18]: No this is fine

File: middleware/cors/cors.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (sixcolors) [reply middleware/cors/cors.go:229]: Isn't this already covered by Test_CORS_Headers_BasedOnRequestType, can't we just improve those tests?

File: middleware/cors/utils.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (sixcolors) [reply middleware/cors/utils.go:71]: we use github.com/stretchr/testify/assert

File: middleware/cors/cors.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (sixcolors) [reply middleware/cors/cors.go:229]: I think you're missing the implications of what "the presence or absence of the Origin and Access-Control-Request-Method headers" implies. If those headers are both present it's a preflight request when called with Options request, of a simple request when called with another method. If either is missing it's not a CORS request.

**Review by ReneWerner87 — APPROVED** [review]

File: middleware/cors/cors.go [hunk 6]
```diff
@@ -13,12 +13,12 @@ type Config struct {
 	// Next defines a function to skip this middleware when returned true.
 	//
 	// Optional. Default: nil
-	Next func(c fiber.Ctx) bool
+	Next func(c *fiber.Ctx) bool
```
Comment (ReneWerner87) [inline middleware/cors/cors.go:16]:
```suggestion
	Next func(c fiber.Ctx) bool
```

**Comment by ReneWerner87** [comment]
@sixcolors pls check my 2 last suggestions

**Pull request merged**